File: 139. Word Break/wordBreak.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        if not s:
            return False
        if not wordDict:
            return False

        n = len(s)
        maxLen = len(wordDict[0])
        for word in wordDict:
            maxLen = max(maxLen, len(word))

        dp = [False] * (n + 1)
        dp[0] = True
        for i in range(1, n + 1):
            for j in range(max(i - maxLen, 0), i):
                if dp[j] and s[j : i] in wordDict:
                    dp[i] = True
                    break

        return dp[n]
        

File: 139. Word Break/wordBreak.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int n = s.length();
        boolean[] dp = new boolean[n + 1];
        
        int maxLen = 0;
        for (String word: wordDict) {
            maxLen = Math.max(maxLen, word.length());
        }

        dp[0] = true;
        for (int i = 1; i <= n; i++) {
            for (int j = Math.max(i - maxLen, 0); j < i; j++) {
                if (dp[j] == true && wordDict.contains(s.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[n];
    }
}